ygen/testdata/proto/proto-test-e.uncompressed.proto-test-e.formatted-txt

```diff
@@ -21,8 +21,8 @@ message AnimalKey {
     SPECIESENUM_OVIS_CANADENSIS = 3;
   }
   oneof species {
-    SpeciesEnum species_speciesenum = 102559808;
-    string species_string = 236397324;
+    SpeciesEnum species_speciesenum = 97881385;
```

Comment:
This change seems to be unrelated to the description of the change - since now these `oneof` fields have changing field numbers. Is there a reason that's the case (it'll cause backwards incompatible protos to be generated).

Reply:
Previously we used the leafref's resolved path (which points to a union) to determine the field numbers. Now we use just the leafref's path to avoid colliding. Since this is backwards incompatible, I'll change it so that only when this collision happens does the behaviour change.

Reply:
I'm missing something here. Sorry.

So we're calling:

```golang
ft, err := fieldTag(fmt.Sprintf("%s_%s", e.Path(), strings.ToLower(tn)))
```

when we generate the `oneof` - in this case `e.Path()` would be the path to the field that defines the leafref, not the leafref's specified `path` argument -- so `/list/one` or `/list/two` if there was a `list` with keys called `one` and `two`. In this case, shouldn't the values be different here since we're hashing with `/list/one_string` and `/list/two_string` for the fields of the union?

What was the clash that was occurring previously?

I'm a little concerned about only doing it when there is a conflict since this has the potential to rely on the ordering of the schema (which we usually don't depend on).

Reply:
The issue occurs when `/list/one` and `/list/two` both point to the same path somewhere else, say `/container/foo`, as exemplified by the new integration test case. This is due to these lines, which calls `resolveLeafrefTarget`, causing both to hash on `/container/foo`, thus creating the collision:
https://github.com/openconfig/ygot/blob/299b0613d66380f125ebb1836392d6e8448bd4d1/ygen/protogen.go#L1170-L1182
The original fix is to use `/list/one` and `/list/two` as you pointed out, but it may introduces a backwards compatibility issue so I then proposed to use `/container/foo` and `/list/two`.

The OP's example has an indirect issue, where the leafref points to a leafref and which collides. This becomes an issue when `compress_paths=true`, because `FindAllChildren` promotes the config/state leaf to the directory.

Reply:
Thanks for the explanation. I'd missed that there was a different context for generating the key protobufs.

I think this fix should be stable, since the `key` field in YANG is ordered and must be kept in the same order, so we'll always hash on `two`. I prefer the option of keeping things backwards compatible, but I wonder if it's counter-intuitive in the future. Should we add a comment that just links to this PR discussion so we remember why?

Thanks again for walking me through the code :-)

Reply:
Np, added comment now with link to this thread.